swagger: add docs for vehicle/stats endpoint

sjain28 opened the pull request:
Adds docs for a `vehicle/id/stats` endpoint that returns engine state and digio (digital aux) objectStats within a timerange (no greater than 1 hour). 

https://paper.dropbox.com/doc/Mini-Dev-Spec-Pike-Electric-API-Updates--AU4cCkqxlRH7SrVwwUjOiCMVAg-wzPIvSmJf8MVofdtJXml6


Dev work on this endpoint is not complete, but Pike wanted to see an API spec.

oliversong commented on swagger.json:
```diff
@@ -1821,6 +1821,57 @@
             }
           }
         },
+        "/fleet/vehicles/{vehicle_id}/stats": {
```
out of curiosity, is there an interface or generator to create this json? or do we have to hand write it?

sjain28 replied:
We have to hand write it at the moment.

katielee-samsara commented on swagger.json:
```diff
@@ -1821,6 +1821,57 @@
             }
           }
         },
+        "/fleet/vehicles/{vehicle_id}/stats": {
```
Nit: the current format is {vehicle_id:[0:9]+} for the other endpoints

sjain28 replied:
I have `{vehicle_id:[0:9]+} ` in the `summary` for the `stats` endpoint that I'm adding. It doesn't look like we add the regex in the list of routes here, see L1765.

katielee-samsara commented on swagger.json:
```diff
@@ -1821,6 +1821,57 @@
+              "get": {
+                  "tags": ["Fleet"],
+                  "summary": "/fleet/vehicles/{vehicle_id:[0-9]+}/stats",
+                  "description": "Fetch engine state and digital aux data for a given vehicle between a start/end time. The maximum query duration is one hour.",
```
Left comment within [mini dev spec](https://paper.dropbox.com/doc/Mini-Dev-Spec-Pike-Electric-API-Updates--AU8GqZ4wvVh5~ma2YfbYZ5u5Ag-wzPIvSmJf8MVofdtJXml6). I'm curious to know if this approach is scalable, and if it would be possible to instead have the developer pass in parameters to request certain object stats.

sjain28 replied:
Added a `filter` argument where caller can request particular stats.

katielee-samsara commented on swagger.json:
```diff
@@ -3105,6 +3156,66 @@
+            "description": "Engine state and digital aux data for a vehicle.",
+            "properties": {
+                "engineState": {
+                    "$ref": "#/definitions/ObjectStatIntSeries"
```
Left comment in mini dev spec. Could you add more information on what is the expected response array size? Are object stats created every x seconds, or do we only show object stats for a state change for these boolean values?

sjain28 replied:
For `engineState`, we only show object stats for a state change.

katielee-samsara replied:
Is this the same for digital aux data?

sjain28 replied:
yes

sjain28 commented on swagger.json:
```diff
@@ -1821,6 +1821,57 @@
+                        "$ref": "#/definitions/VehicleStats"
+                      }
+                    },
+                    "default": {
```
In this case, `default` describes the schema of the API response for any code other than 200. In the `responses` object we can enumerate response types for different status codes, and have a  `default` option to handle all the remaining status codes.

sjain28 commented on swagger.json:
```diff
@@ -3105,6 +3156,66 @@
                 }
             }
         },
+        "VehicleStats": {
```
No formal order, but yeah. I can move the object stats stuff to the bottom so that they are grouped together.

sjain28 replied:
Actually I'd prefer to keep them where they are, grouped with `VehicleLocation` since that response type also contains objectstats.

sjain28 commented on swagger.json:
```diff
@@ -3105,6 +3156,66 @@
+                }
+            }       
+        },
+        "ObjectStatIntSeries": {
```
Weird, looks fine here. 

<img width="690" alt="screen shot 2019-01-04 at 10 38 20 am" src="https://user-images.githubusercontent.com/9773036/50704621-e5a5a180-100c-11e9-90b4-6686288487a0.png">

sjain28 commented:
@katielee-samsara I tweaked some of the language and addressed your comments above and on the dev spec.

sjain28 commented:
@silverspace made the following changes after our conversation: 
* Changed route to `/vehicles/stats` so that we return stats for all vehicles
* Added `filter` argument
* Changed "digital aux" to "digital input" to match cloud dashboard language

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                },
+                {
+                  "name": "endMs",
+                  "description": "Time in Unix epoch milliseconds for the end of the query (cannot exceed 1 hour)",
```
Same as above. Would we always return the most recent timeseries value if we passed graphql (nowMs, nowMs) as the (startMs, endMs) values? If so, that would be an ideal default.

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                {
+                  "name": "startMs",
+                  "description": "Time in Unix epoch milliseconds for the start of the query (cannot exceed 1 hour)",
+                  "required": true,
```
Do we want to make startMs and endMs required params? For instance, we could default endMs to `now`.

sjain28 replied:
Both are required in `/fleet/vehicles/locations`, and I'd prefer to have `/stats` match that endpoint.

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                  "format": "int64"
+                },
+                { 
+                  "name": "filter",
```
Sorry for changing the naming convention here -- could we rename this to `series`?

sjain28 replied:
Renamed

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                },
+                { 
+                  "name": "filter",
+                  "description": "Comma-separated list of stat types. Options are engineState, auxInput1, and auxInput2.",
```
Should we just have a single `aux-inputs` type that returns both `auxInput1` and `auxInput2`?

silverspace replied:
I think this should be a required query parameter, since we want to avoid returning all timeseries data if this API gets larger.

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                  "$ref": "#/parameters/accessTokenParam"
+                },
+                {
+                  "name": "startMs",
```
These parameter should all be in the "get" method section.

sjain28 replied:
Moved the parameters.

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                    "200": {
+                      "description": "Returns engine state and aux input data for all vehicles in the group between a start/end time, with a maximum query duration of one hour.",
+                      "schema": {
+                        "type": "array",
```
Could we return a dictionary at the top level rather than an array? That way the result type is extensible if we want to add additional metadata in the future.

sjain28 replied:
Changed the return type

silverspace commented on swagger.json:
```diff
@@ -3105,6 +3168,76 @@
+                },
+                "value": {
+                    "type": "integer",
+                    "description": "Integer value.",
```
Nit: This description is not helpful, could we add some more description about what this value represents for different aux types?

silverspace commented on swagger.json:
```diff
@@ -3105,6 +3168,76 @@
+        },
+        "ObjectStatEngineState": {
+            "type": "object",
+            "description": "The state of the vehicle over time. 0 = Off, 1 = Running, 2 = Idle.",
```
Could this be a swagger enum? https://swagger.io/docs/specification/2-0/enums/

Could we also use the human readable name rather than the int value?

sjain28 replied:
Yep, changed to be an enum of human-readable values.

silverspace commented on swagger.json:
```diff
@@ -3105,6 +3168,76 @@
+                    "properties": {
+                        "name": {
+                            "type": "string",
+                            "example": "Boom"
```
Could we enumerate the DIGIO input types here? https://swagger.io/docs/specification/2-0/enums/

silverspace commented on swagger.json:
```diff
@@ -1821,6 +1821,69 @@
+                                    "format": "int64",
+                                    "example": "33"
+                                },
+                                "stats": {
```
Is this stats wrapper required, or should we directly inline the relevant series here?

sjain28 replied:
Inlined the timeseries objects.

oliversong commented on swagger.json:
```diff
@@ -1821,6 +1821,80 @@
+                      "type": "integer"
+                    },
+                    { 
+                      "name": "series",
```
should this be `filter` ?

sjain28 replied:
Renamed it to `series` as per Greg's comment earlier.

oliversong replied:
or, should the example be `series=engineState,auxInput2`?

sjain28 replied:
changed the example
